File: Day-37/BookAllocationProblem.js
function isPossible(arr, students, mid) {
  let studentCount = 1;
  let pageSum = 0;

  for (let i = 0; i < arr.length; i++) {
    if (pageSum + arr[i] <= mid) {
      pageSum += arr[i];
    } else {
      studentCount++;
      if (studentCount > students || arr[i] > mid) {
        return false;
      } else {
        pageSum = arr[i];
      }
    }
  }
  return true;
}

function bookAllocation(arr, students) {
  let sum = 0;
  for (let i = 0; i < arr.length; i++) {
    sum += arr[i];
  }
  let start = 0;
  let end = sum;
  let mid = parseInt((start + end) / 2);
  let ans = -1;

  while (start < end) {
    if (isPossible(arr, students, mid)) {
      ans = mid;
      end = mid - 1;
    } else {
      start = mid + 1;
    }
    mid = parseInt((start + end) / 2);
  }
  return ans;
}

console.log(bookAllocation([10, 20, 30, 40], 2));

File: Day-38/BookAllocationProblem.py
def isPossible(arr, students, mid):
    studentCount = 1
    pageSum = 0
    for i in range(0, len(arr)):
        if(pageSum + arr[i] <= mid):
            pageSum += arr[i]
        else:
            studentCount += 1
            if(studentCount > students or arr[i]>mid):
                return False
            else:
                pageSum = arr[i]
    return True


def BookAllocationProblem(arr, students):
    sum =0
    for i in range(1, len(arr)):
        sum += arr[i]
    start = arr[0]
    end = sum
    mid = (start+end)//2
    ans = -1
    while(start<end):
        if(isPossible(arr, students, mid)):
            ans = mid
            end = mid -1
        else:
            start = mid +1

        mid = (start+end )//2

    return ans
    



print(BookAllocationProblem([10,20,30,40], 2))